### PR TITLE
perlPackages.SQLTranslator: fix some scripts issues

### PR DESCRIPTION
**Pull request by pacien**
###### Motivation for this change

* perlPackages.SQLTranslator: shorten long shebang 
The long shebang was causing troubles for running the Perl scripts on
both Darwin and Linux. This seems to solve the issue on Perl 5.32,
but the scripts remain broken on Perl 5.34.

* perlPackages.SQLTranslator: add missing dep
This adds a missing dependency for the `sqlt-graph` script provided in
the SQLTranslator package.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131414` at 60d8c34c run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>perl532Packages.DBICxTestDatabase</li>
    <li>perl532Packages.DBIxClassDynamicDefault</li>
    <li>perl532Packages.SQLTranslator</li>
    <li>perl534Packages.DBICxTestDatabase</li>
    <li>perl534Packages.DBIxClassDynamicDefault</li>
    <li>perl534Packages.SQLTranslator</li>
  </ul>
</details>

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -18805,8 +18805,14 @@ let
       url = "mirror://cpan/authors/id/I/IL/ILMARI/SQL-Translator-1.62.tar.gz";
       sha256 = "096fq62iphmxgcfcayg4i0wf09y1rl1ar6hrcxfqsbrsmkwlzk8a";
     };
-    buildInputs = [ FileShareDirInstall JSONMaybeXS TestDifferences TestException XMLWriter YAML ];
-    propagatedBuildInputs = [ CarpClan DBI FileShareDir Moo PackageVariant ParseRecDescent TryTiny ];
+    buildInputs = [ shortenPerlShebang FileShareDirInstall JSONMaybeXS TestDifferences TestException XMLWriter YAML ];
```
[the remark this answers is not in the document]
Comment (pacien) [reply pkgs/top-level/perl-packages.nix]: Fixed and amended :+1:

[the remark this answers is not in the document]
Comment (pacien) [reply pkgs/top-level/perl-packages.nix]: Woops, amended to the wrong commit... Fixing...

**Comment by stigtsp** [comment]
The shebangs in the scripts are set to `#!/nix/store/.....coreutils-x.xx/bin/env perl` this might be the reason why they appear to fail on perl 5.34.

Adding `patchShebangs` might resolve the problem you're seeing.

```
    postPatch = ''
      patchShebangs bin
    '';
```

PS: Try it without `shortenPerlshebang`, it might only be needed on Darwin.

**Comment by pacien** [comment]
On Linux, with only `patchShebangs`, without `shortenPerlShebang`:

```
❯ nix build .#perl534Packages.SQLTranslator
❯ ./result/bin/sqlt-diagram --help
/nix/store/nvlnl2swxmk50x25mzn71rhc0gzpylfr-coreutils-8.32/bin/env: ‘perl -I/nix/store/y29hhfh2gjmiagdanzywabjbd2y4g79c-perl-5.34.0/lib/perl5/site_perl -I/nix/store/03af9jbrywvrkd1vcjv6hcna4qlx4aw3-perl5.34.0-File-ShareDir-Install-0.13/lib/perl5/site_perl’: No such file or directory
/nix/store/nvlnl2swxmk50x25mzn71rhc0gzpylfr-coreutils-8.32/bin/env: use -[v]S to pass options in shebang lines
```

On Linux, with both `patchShebangs` and `shortenPerlShebang`:

```
❯ nix build .#perl534Packages.SQLTranslator
❯ ./result/bin/sqlt-diagram --help
Can't locate SQL/Translator.pm in @INC (you may need to install the SQL::Translator module) (@INC contains: /nix/store/y29hhfh2gjmiagdanzywabjbd2y4g79c-perl-5.34.0/lib/perl5/site_perl /nix/store/03af9jbrywvrkd1vcjv6hcna4qlx4aw3-perl5.34.0-File-ShareDir-Install-0.13/lib/perl5/site_perl /nix/store/pssx0kzw5cfvf8ph9fbwsywx6jwda2cr-perl5.34.0-JSON-MaybeXS-1.004003/lib/perl5/site_perl /nix/store/ixbhwbwfwkhja1705mdy44pbq2z24wg2-perl5.34.0-Test-Differences-0.67/lib/perl5/site_perl /nix/store/mzix90bpyscj2mp5y5wwi2klysw0a4yp-perl5.34.0-Capture-Tiny-0.48/lib/perl5/site_perl /nix/store/n0lkl84czhgqc7b44ib33szkyh9slgcb-perl5.34.0-Text-Diff-1.45/lib/perl5/site_perl /nix/store/9xx0i68ipw55byjy1d57s6fyx16cyp02-perl5.34.0-Algorithm-Diff-1.1903/lib/perl5/site_perl /nix/store/9ndic1b2v9104f4hxxp72mh37cmkqk31-perl5.34.0-Test-Exception-0.43/lib/perl5/site_perl /nix/store/giii2g0n31s5q7ji046jxxxfzg7f3v0l-perl5.34.0-Sub-Uplevel-0.2800/lib/perl5/site_perl /nix/store/mspl4xlznw7mz7m66xkxrk1y67if696f-perl5.34.0-XML-Writer-0.900/lib/perl5/site_perl /nix/store/fppqx5am25qfc730q91iw0qyzbnf0v05-perl5.34.0-YAML-1.30/lib/perl5/site_perl /nix/store/2x9m86yfm2k15byn7lg3f2q8ii8h8mi3-perl5.34.0-Carp-Clan-6.08/lib/perl5/site_perl /nix/store/y9r3zcgsl108xn5m8dd4fmg48pv0ymkh-perl5.34.0-DBI-1.643/lib/perl5/site_perl /nix/store/dkn0xhk61vbn6hpaqcs7p10h5y3ky707-perl5.34.0-File-ShareDir-1.118/lib/perl5/site_perl /nix/store/7hb7n294k4fwsjmf2vrqac0zsdcy9haz-perl5.34.0-Class-Inspector-1.36/lib/perl5/site_perl /nix/store/67g6lxisvx13rzz860yaikl79iplf1f8-perl5.34.0-Moo-2.004004/lib/perl5/site_perl /nix/store/2lmclwbwrz68x6d6dzrw4i3qjqvliprn-perl5.34.0-Class-Method-Modifiers-2.13/lib/perl5/site_perl /nix/store/2y3q96wcg2n6zcx9imvmapzp94qz41kb-perl5.34.0-Module-Runtime-0.016/lib/perl5/site_perl /nix/store/kp1104m3lcgnln3r1rz52akm0pf1vw9i-perl5.34.0-Role-Tiny-2.001004/lib/perl5/site_perl /nix/store/7azh3vjcmkdi6v52l88hh0hifz2myfkg-perl5.34.0-Sub-Quote-2.006006/lib/perl5/site_perl /nix/store/8ia7kyw5cfbbg814idya82r53pdlia5v-perl5.34.0-Package-Variant-1.003002/lib/perl5/site_perl /nix/store/fl84lfrvgsdrzm5xwbvh186y2bgy4r7d-perl5.34.0-Import-Into-1.002005/lib/perl5/site_perl /nix/store/8hwhaw5hsna4zmnq2vga2yfkm8hyzc6l-perl5.34.0-strictures-2.000006/lib/perl5/site_perl /nix/store/yajrawq7xxaxkpl8m8h3pzwkal9lmlx4-perl5.34.0-Parse-RecDescent-1.967015/lib/perl5/site_perl /nix/store/qjiynfaib7r941bap6jy93vbgampwwcs-perl5.34.0-Try-Tiny-0.30/lib/perl5/site_perl /nix/store/359i4xgm74hvl5rwhh4brp4356h91lqn-perl5.34.0-GraphViz-2.24/lib/perl5/site_perl /nix/store/j5dd7pifpqjxf812cm462j5gna9x8b5f-perl5.34.0-File-Which-1.23/lib/perl5/site_perl /nix/store/b325k3cw8kig2wvnpw4zk234zb6fska1-perl5.34.0-IPC-Run-20200505.0/lib/perl5/site_perl /nix/store/9b3ghpfwh38aqf042a0l04n3vdf9vazx-perl5.34.0-IO-Tty-1.15/lib/perl5/site_perl /nix/store/p6ajchakms9rm4j62g4yfb3aqh3zqssq-perl5.34.0-XML-Twig-3.52/lib/perl5/site_perl /nix/store/9nr4drhi9110a544ag6xbijbz6w9bby6-perl5.34.0-XML-Parser-2.46/lib/perl5/site_perl /nix/store/rk7q85mryrwv0qcyx10qvy2qx0s76xyx-perl5.34.0-libwww-perl-6.49/lib/perl5/site_perl /nix/store/m292qqzzzaj9hrmpidl6972zsaavqbqi-perl5.34.0-File-Listing-6.14/lib/perl5/site_perl /nix/store/37njap8i5mmflzyp4msgkmqwb6rlj1sq-perl5.34.0-HTTP-Date-6.05/lib/perl5/site_perl /nix/store/hrh82rg7fpxa85dlp9l3llm71pzay0dj-perl5.34.0-TimeDate-2.33/lib/perl5/site_perl /nix/store/rq5xzi0rsp3q8v6kqdw806pb0gzfr35f-perl5.34.0-HTML-Parser-3.75/lib/perl5/site_perl /nix/store/14nx9wmizvsl024v65r2zw3fsv8a5aga-perl5.34.0-HTML-Tagset-3.20/lib/perl5/site_perl /nix/store/h28y7ainrfg5nlxnwj96g3amnbmiy610-perl5.34.0-HTTP-Message-6.26/lib/perl5/site_perl /nix/store/80yqyx7vnxa3par83lvrw0njw4dx3d36-perl5.34.0-Encode-Locale-1.05/lib/perl5/site_perl /nix/store/y7fahdp1kbh7s0bk7vcmyhjrvlv1bk0x-perl5.34.0-IO-HTML-1.004/lib/perl5/site_perl /nix/store/3nwb37n7ffjvw8bcv3z63alz7xfggn28-perl5.34.0-LWP-MediaTypes-6.04/lib/perl5/site_perl /nix/store/w0a7vcg03ihgv3cxclp3qjij59ixwy57-perl5.34.0-URI-5.05/lib/perl5/site_perl /nix/store/jgjvdnswg8k6xbq3illk8m38103jlb6w-perl5.34.0-HTTP-Cookies-6.09/lib/perl5/site_perl /nix/store/qixgmaa7kfrdgdap5m8mqgplisy43w08-perl5.34.0-HTTP-Daemon-6.01/lib/perl5/site_perl /nix/store/8s0b48rpwvaidn52x0vp2vpfqr2qj5s1-perl5.34.0-HTTP-Negotiate-6.01/lib/perl5/site_perl /nix/store/5834z3dbsbmjcr9crfk8z64rpxp0svp3-perl5.34.0-Net-HTTP-6.19/lib/perl5/site_perl /nix/store/bb2hkalggi4v5ckcammf44751msvn2vj-perl5.34.0-WWW-RobotRules-6.02/lib/perl5/site_perl /nix/store/jpnj50b18cs5hmbgffmdyjpg4al0xfv4-perl5.34.0-XML-XPath-1.44/lib/perl5/site_perl /nix/store/vcyz8c5cmplasy1hl5wj12n9ks5h8hx8-perl5.34.0-GD-2.73/lib/perl5/site_perl /nix/store/hb9pi1awa49spy6nj3r1wnmyv2yymcll-perl5.34.0-SQL-Translator-1.62/lib/perl5/site_perl /nix/store/n7hbdyp3bsmdxy2lcxivaxnq4nv8ndv3-perl-5.32.1/lib/perl5/site_perl/5.32.1/x86_64-linux-thread-multi /nix/store/n7hbdyp3bsmdxy2lcxivaxnq4nv8ndv3-perl-5.32.1/lib/perl5/site_perl/5.32.1 /nix/store/n7hbdyp3bsmdxy2lcxivaxnq4nv8ndv3-perl-5.32.1/lib/perl5/5.32.1/x86_64-linux-thread-multi /nix/store/n7hbdyp3bsmdxy2lcxivaxnq4nv8ndv3-perl-5.32.1/lib/perl5/5.32.1) at ./result/bin/sqlt-diagram line 78.
BEGIN failed--compilation aborted at ./result/bin/sqlt-diagram line 78.
```

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -18805,8 +18805,15 @@ let
       url = "mirror://cpan/authors/id/I/IL/ILMARI/SQL-Translator-1.62.tar.gz";
       sha256 = "096fq62iphmxgcfcayg4i0wf09y1rl1ar6hrcxfqsbrsmkwlzk8a";
     };
+    nativeBuildInputs = [ shortenPerlShebang ];
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:18808]:
```suggestion
    nativeBuildInputs = lib.optional stdenv.isDarwin shortenPerlShebang;
```

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -18805,8 +18805,15 @@ let
...
     buildInputs = [ FileShareDirInstall JSONMaybeXS TestDifferences TestException XMLWriter YAML ];
-    propagatedBuildInputs = [ CarpClan DBI FileShareDir Moo PackageVariant ParseRecDescent TryTiny ];
+    propagatedBuildInputs = [ CarpClan DBI FileShareDir Moo PackageVariant ParseRecDescent TryTiny GraphViz GD ];
+    # shebang include path too long for both Linux and Darwin
+    postInstall = ''
+      for file in $out/bin/*; do
+        shortenPerlShebang $file
+      done
+    '';
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:18816]:
```suggestion

    postPatch = ''
      patchShebangs script
    '';

    postInstall = lib.optionalString stdenv.isDarwin ''
      for file in $out/bin/*; do
        shortenPerlShebang $file
      done
    '';
```

**Review by stigtsp — COMMENTED** [review]
Ah, during the patch phase the scripts are in the `script/` subdirectory, not `bin/`. The following code appears to work for me (not tested on Darwin though)

File: pkgs/top-level/perl-packages.nix (hunk 3 above)
Comment (pacien) [reply pkgs/top-level/perl-packages.nix]: Thanks, this works; Amended!

File: pkgs/top-level/perl-packages.nix (hunk 2 above)
Comment (pacien) [reply pkgs/top-level/perl-packages.nix]: Thanks, this works; Amended!

**Comment by stigtsp** [comment]
@GrahamcOfBorg build perl532Packages.SQLTranslator perl534Packages.SQLTranslator

**Review by stigtsp — APPROVED** [review]
Looks good to me, thanks for the contribution

This change adds `GD` and `GraphViz` as dependencies to this module, increasing size a bit.

```
# Before
/nix/store/bi2gmayqi56vm3yqzxr31qnmagsagdpq-perl5.34.0-SQL-Translator-1.62	   3.2M	  92.5M
# After
/nix/store/h48yyy3yfd0hyyxz4q3zwzq5nf8pwvi6-perl5.34.0-SQL-Translator-1.62         3.2M  111.3M
```

I've had a look at what other distros do: They are dependencies in [Fedora](https://src.fedoraproject.org/rpms/perl-SQL-Translator/blob/f34/f/perl-SQL-Translator.spec), recommended dependencies in [Debian](https://packages.debian.org/sid/libsql-translator-perl), but are not included in [Alpine](https://pkgs.alpinelinux.org/package/edge/community/x86/perl-sql-translator). Seems reasonable to add these dependencies.

**Pull request merged**